### ISSUE TITLE
Large Confidence Intervals in Variational Multi-Task setup

### ISSUE THREAD

**Issue by AndreaBraschi** [cornellius-gp/gpytorch#1959]
Hi,

I've been trying to use the intrinsic coregionalization model as a variational strategy to model correlation between tasks Here is the code that I've written:

``` 
class ICMVariationalStrategy_single(_VariationalStrategy):

    def __init__(
        self,
        base_variational_strategy: gpytorch.variational.VariationalStrategy,
        num_tasks: int,
        base_kernel: gpytorch.kernels.Kernel,
        rank: int
    ):
        Module.__init__(self)
        self.base_variational_strategy = base_variational_strategy

        # kernel modules
        self.base_kernel = base_kernel
        self.covar_module = gpytorch.kernels.MultitaskKernel(
            self.base_kernel, num_tasks=num_tasks, rank=rank
        )
    @property
    def prior_distribution(self):
        return self.base_variational_strategy.prior_distribution

    @property
    def variational_distribution(self):
        return self.base_variational_strategy.variational_distribution

    @property
    def variational_params_initialized(self):
        return self.base_variational_strategy.variational_params_initialized

    def kl_divergence(self):
        return super().kl_divergence().sum(dim=-1)

    def __call__(self, x, prior=False, **kwargs):

        # Get an approximate posterior for each task
        q_f = self.base_variational_strategy(x)
        # Mean
        mean = q_f.mean.transpose(-1, -2)

        # Covariance
        covar_x = self.covar_module(x)

        # Distribution
        distribution = gpytorch.distributions.MultitaskMultivariateNormal(mean, covar_x)
        return distribution 
```

Here is the code where I use the above class in the tutorial 'Variational GPs w/ Multiple Outputs'. Basically, I want to try to use the ICM instead of the lcm that was shown in the tutorial.
```
import numpy as np
import pandas as pnd
import torch
import gpytorch
import os.path
from matplotlib import pyplot as plt
from tqdm.notebook import trange
import scipy
import math
import OpenSim_utils


from gpytorch.models import ExactGP, ApproximateGP
from gpytorch.means import ConstantMean, MultitaskMean, ZeroMean
from gpytorch.kernels import MultitaskKernel, RBFKernel, ScaleKernel
from gpytorch.distributions import MultitaskMultivariateNormal
from gpytorch.likelihoods import MultitaskGaussianLikelihood
from gpytorch.variational import TrilNaturalVariationalDistribution
from gpytorch.variational.variational_strategy import *
import sys



class Approximate_MultitaskGP(ApproximateGP):
    def __init__(self, num_tasks: int, n_inducing: int, x: torch.Tensor):
        self.num_tasks = num_tasks
        self.x = x

        inducing_inputs = torch.randn(n_inducing, 1)
        self.inducing_inputs = inducing_inputs.expand(*torch.Size([self.num_tasks]), *inducing_inputs.shape[-2:])

        variational_dist = TrilNaturalVariationalDistribution(n_inducing, batch_shape=torch.Size([self.num_tasks]))

        q_f = ICMVariationalStrategy_single(VariationalStrategy(self, self.inducing_inputs, variational_dist, learn_inducing_locations=True),
                                                self.num_tasks, RBFKernel(), rank=1)

        super().__init__(q_f)

        self.mean_module = ZeroMean()
        self.covar_module = q_f.base_kernel

    def forward(self, x):
        mean = self.mean_module(x)
        covar = self.covar_module(x)
        dist = MultivariateNormal(mean, covar)
        return dist

train_x = torch.linspace(0, 1, 100)

y = torch.stack([
    torch.sin(train_x * (2 * math.pi)) + torch.randn(train_x.size()) * 0.2,
    torch.cos(train_x * (2 * math.pi)) + torch.randn(train_x.size()) * 0.2,
    torch.sin(train_x * (2 * math.pi)) + 2 * torch.cos(train_x * (2 * math.pi)) + torch.randn(train_x.size()) * 0.2,
    -torch.cos(train_x * (2 * math.pi)) + torch.randn(train_x.size()) * 0.2,
], -1)


n_inducing = 16
num_tasks = y.shape[-1]
Model = Approximate_MultitaskGP(num_tasks, n_inducing, train_x)
Likelihood = MultitaskGaussianLikelihood(num_tasks=num_tasks)

# set training mode
Model.train()
Likelihood.train()

# NGD and Adam optimisers
NGD = gpytorch.optim.NGD(Model.variational_parameters(), num_data=round(y.shape[-2] * 0.5), lr=0.01)
Adam = torch.optim.Adam([
    {'params': Model.hyperparameters()},
    {'params': Likelihood.parameters()}],
     lr=0.001)

# list of variational parameters
variational_param = []
for name, param in Model.named_variational_parameters():
    variational_param.append(name)

# list of hyperparameters
hyperparameters = []
for name, param in Model.named_hyperparameters():
    hyperparameters.append(name)

# Objective function
mll = gpytorch.mlls.VariationalELBO(Likelihood, Model, num_data=y.shape[-2])

iterator = trange(5000, leave=True)
training_iter = 5000
loss_list = []
kt_list = np.zeros((training_iter, 1))

for i in iterator:
    NGD.zero_grad()
    Adam.zero_grad()
    output = Model(train_x)
    loss = -mll(output, y)
    loss_list.append(loss.item())
    loss.backward()
    print('Iter %d/%d - Loss: %.3f' % (i + 1, training_iter, loss.item()))

    NGD.step()
    Adam.step()

print(Model.variational_strategy.covar_module.task_covar_module.covar_factor)
# Set into eval mode
Model.eval()
Likelihood.eval()

# Prediction
with torch.no_grad(), gpytorch.settings.fast_pred_var():
    test_x = torch.linspace(0, 1, 51)
    predictions = Likelihood(Model(test_x))
    mean = predictions.mean
    mean_np = mean.detach().numpy()
    lower, upper = predictions.confidence_region()
    lower_np = lower.detach().numpy()
    upper_np = upper.detach().numpy()

# Initialize plots
fig, axs = plt.subplots(1, num_tasks, figsize=(4 * num_tasks, 3))

for task, ax in enumerate(axs):
    # training data
    ax.plot(train_x.detach().numpy(), y[:, task].detach().numpy(), 'k*')
    # Prediction
    ax.plot(test_x.detach().numpy(), mean[:, task].detach().numpy(), 'b')
    # confidence intervals
    ax.fill_between(test_x.detach().numpy(), lower[:, task].detach().numpy(), upper[:, task].detach().numpy(),
                    alpha=0.5)

    ax.legend(['Training Data', 'Prediction', 'CI'])
    ax.set_title(f'Task {task + 1}')

plt.show()
```

After manually trying different learning rates for the NGD and the Adam optimisers, I found that 0.01 and 0.001 is what gives the best numerical stability in the optimisation. However, numerous iterations are needed for such a low learning rate in the Adam optimiser, otherwise the confidence intervals are very large. (you can try to lower the number of iterations and see how the confidence intervals get larger in the graphs).

However, the variational strategy that I showed doesn't seem to work very well with my data. Predictions aren't very accurate and confidence intervals are massively large.

I was wondering if there is something apparently wrong with the code that I'm using and if anyone notices something that might be fixed.

Any help would be very much appreciated.

Cheers,

Andrea
